README.md
# express-basic-auth

[![npm version](https://badge.fury.io/js/%40blossomfinance%2Fexpress-basic-auth.svg)](https://badge.fury.io/js/%40blossomfinance%2Fexpress-basic-auth)
![npm (scoped)](https://img.shields.io/npm/v/@blossomfinance/express-basic-auth)
[![CircleCI](https://circleci.com/gh/matmar10/express-basic-auth/tree/master.svg?style=svg)](https://circleci.com/gh/matmar10/express-basic-auth/tree/master)
[![David](https://img.shields.io/david/strongloop/express.svg)]()
![TypeScript compatible](https://img.shields.io/badge/typescript-compatible-brightgreen.svg)
[![MIT Licence](https://badges.frapsoft.com/os/mit/mit.svg?v=103)](https://opensource.org/licenses/mit-license.php)

Simple plug & play HTTP basic auth middleware for Express.

## How to install

Just run

```shell
npm install express-basic-auth
```

## How to use

The module will export a function, that you can call with an options object to
get the middleware:

```js
const app = require('express')()
const basicAuth = require('express-basic-auth')

app.use(basicAuth({
    users: { 'admin': 'supersecret' }
}))
```

The middleware will now check incoming requests to match the credentials
`admin:supersecret`.

The middleware will check incoming requests for a basic auth (`Authorization`)
header, parse it and check if the credentials are legit. If there are any
credentials, an `auth` property will be added to the request, containing
an object with `user` and `password` properties, filled with the credentials,
no matter if they are legit or not.

**If a request is found to not be authorized**, it will respond with HTTP 401
and a configurable body (default empty).

### Static Users

If you simply want to check basic auth against one or multiple static credentials,
you can pass those credentials in the `users` option:

```js
app.use(basicAuth({
    users: {
        'admin': 'supersecret',
        'adam': 'password1234',
        'eve': 'asdfghjkl',
    }
}))
```

The middleware will check incoming requests to have a basic auth header matching
one of the three passed credentials.

### Custom authorization

Alternatively, you can pass your own `authorizer` function, to check the credentials
however you want. It will be called with a username and password and is expected to
return `true` or `false` to indicate that the credentials were approved or not.

When using your own `authorizer`, make sure **not to use standard string comparison (`==` / `===`)**
when comparing user input with secret credentials, as that would make you vulnerable against
[timing attacks](https://en.wikipedia.org/wiki/Timing_attack). Use the provided `safeCompare`
function instead - always provide the user input as its first argument. Also make sure to use bitwise
logic operators (`|` and `&`) instead of the standard ones (`||` and `&&`) for the same reason, as
the standard ones use shortcuts.

```js
app.use(basicAuth( { authorizer: myAuthorizer } ))

function myAuthorizer(username, password) {
    const userMatches = basicAuth.safeCompare(username, 'customuser')
    const passwordMatches = basicAuth.safeCompare(password, 'custompassword')

    return userMatches & passwordMatches
}
```

This will authorize all requests with the credentials 'customuser:custompassword'.
In an actual application you would likely look up some data instead ;-) You can do whatever you
want in custom authorizers, just return `true` or `false` in the end and stay aware of timing
attacks.

### Custom Async Authorization

Note that the `authorizer` function above is expected to be synchronous. This is
the default behavior, you can pass `authorizeAsync: true` in the options object to indicate
that your authorizer is asynchronous. In this case it will be passed a callback
as the third parameter, which is expected to be called by standard node convention
with an error and a boolean to indicate if the credentials have been approved or not.
Let's look at the same authorizer again, but this time asynchronous:

```js
app.use(basicAuth({
    authorizer: myAsyncAuthorizer,
    authorizeAsync: true,
}))

function myAsyncAuthorizer(username, password, cb) {
    if (username.startsWith('A') & password.startsWith('secret'))
        return cb(null, true)
    else
        return cb(null, false)
}
```

### Unauthorized Response Body

Per default, the response body for unauthorized responses will be empty. It can
be configured using the `unauthorizedResponse` option.

You can pass one of the following:

- a static value (string or object)
- a function that synchronously returns a value (string or object)
- a middleware

If a string value is returned or produced by the function, it will be
sent as-is as the response along with a 401 status code. Ohterwise,
it will be sent as json via the `.json` method.


```js

// static value
app.use(basicAuth({
    users: { 'Foo': 'bar' },
    unauthorizedResponse: 'Sorry, you are not authorized.'
}))

// a function that synchronously returns a value
app.use(basicAuth({
    users: { 'Foo': 'bar' },
    unauthorizedResponse: function (req) {
        return req.auth
            ? ('Credentials ' + req.auth.user + ':' + req.auth.password + ' rejected')
            : 'No credentials provided'
    }
}))

// a middleware
app.use(basicAuth({
    users: { 'Foo': 'bar' },
    unauthorizedResponse: function (req, res, next) {
      if (req.auth) {
          next(new Error('Login required'));
          return;
      }
      // or send an html page
      if (req.accepts('html') {
          res.render('unauthorized');
          return;
      }
      // or whatever you like!
      res.status(403)
          .set('X-Toynbee-Idea', 'In Kubrick’s 2001 Resurrect Dead On Planet Jupiter')
          .send({
            message: 'Future authorization attempts are futile.'
          });
    }
}))

```

### Challenge

Per default the middleware will not add a `WWW-Authenticate` challenge header to
responses of unauthorized requests. You can enable that by adding `challenge: true`
to the options object. This will cause most browsers to show a popup to enter
credentials on unauthorized responses. You can set the realm (the realm
identifies the system to authenticate against and can be used by clients to save
credentials) of the challenge by passing a static string or a function that gets
passed the request object and is expected to return the challenge:

```js
app.use(basicAuth({
    users: { 'someuser': 'somepassword' },
    challenge: true,
    realm: 'Imb4T3st4pp',
}))
```

## Try it

The repository contains an `example.js` that you can run to play around and try
the middleware. To use it just put it somewhere (or leave it where it is), run

```shell
npm install express express-basic-auth
node example.js
```

This will start a small express server listening at port 8080. Just look at the file,
try out the requests and play around with the options.

## TypeScript usage

A declaration file is bundled with the library. You don't have to install a `@types/` package.

```typescript
import * as basicAuth from 'express-basic-auth'
```

:bulb: **Using `req.auth`**

express-basic-auth sets `req.auth` to an object containing the authorized credentials like `{ user: 'admin', password: 'supersecret' }`.

In order to use that `req.auth` property in TypeScript without an unknown property error, use covariance to downcast the request type:

```typescript
app.use(basicAuth(options), (req: basicAuth.IBasicAuthedRequest, res, next) => {
    res.end(`Welcome ${req.auth.user} (your password is ${req.auth.password})`)
    next()
})
```

:bulb: **A note about type inference on synchronous authorizers**

Due to some TypeScript's type-system limitation, the arguments' type of the synchronous authorizers are not inferred.
For example, on an asynchronous authorizer, the three arguments are correctly inferred:

```typescript
basicAuth({
    authorizeAsync: true,
    authorizer: (user, password, authorize) => authorize(null, password == 'secret'),
})
```

However, on a synchronous authorizer, you'll have to type the arguments yourself:

```typescript
basicAuth({
    authorizer: (user: string, password: string) => (password == 'secret')
})
```

## Tests

The cases in the `example.js` are also used for automated testing. So if you want  
to contribute or just make sure that the package still works, simply run:

```shell
npm test
```


CHANGELOG.md
### [1.3.3](https://github.com/matmar10/express-basic-auth/compare/v1.3.2...v1.3.3) (2021-08-07)


### Continuous Integration

* remove typescript compiler check from build test ([2579b12](https://github.com/matmar10/express-basic-auth/commit/2579b125f2ccf376b32cd04376a1b11f48761378))


### Documentation

* update badges to new forked repo ([3ac09ed](https://github.com/matmar10/express-basic-auth/commit/3ac09ed68d12177edc14b257bfb7756d3256fad3))

### [1.3.2](https://github.com/matmar10/express-basic-auth/compare/v1.3.1...v1.3.2) (2021-08-07)


### Documentation

* point to fork & add contributors ([72bf15b](https://github.com/matmar10/express-basic-auth/commit/72bf15b58e76f65cda015c0faca996067ff2f89c))

### [1.3.1](https://github.com/LionC/express-basic-auth/compare/v1.3.0...v1.3.1) (2021-08-07)

## [1.3.0](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2021-08-07)


### Features

* enable custom unauthorized handling via middleware ([8d3c2e1](https://github.com/LionC/express-basic-auth/commit/8d3c2e1b32131daee6686c8922de86fdbd91d01d)), closes [#23](https://github.com/LionC/express-basic-auth/issues/23)


### Documentation

* add release script using conventional commit changelog ([e13b2d6](https://github.com/LionC/express-basic-auth/commit/e13b2d623fb31636ff029b6ba1613294aa634044))

## [1.2.0](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2019-04-22)

### [1.1.7](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2019-04-22)

### [1.1.6](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2018-10-26)

### [1.1.5](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2018-04-23)

### [1.1.4](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2018-02-09)

### [1.1.3](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2017-10-23)

### [1.1.2](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2017-08-15)

### [1.1.1](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2017-05-29)

## [1.1.0](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2017-05-29)

### [1.0.2](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2017-04-27)

### [1.0.1](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2017-03-18)

## [1.0.0](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2017-02-16)

### [0.3.3](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2017-02-16)

### [0.3.2](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2017-02-14)

### [0.3.1](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2017-02-14)

## [0.3.0](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2017-02-14)

### [0.2.3](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2016-11-07)

### [0.2.2](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2016-10-19)

### [0.2.1](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2016-10-18)

## [0.2.0](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2016-10-18)

### [0.1.3](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2016-10-15)

### [0.1.2](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2016-10-11)

### [0.1.1](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2016-10-11)

## [0.1.0](https://github.com/LionC/express-basic-auth/compare/v1.2.0...v1.3.0) (2016-10-11)



test.js
const should = require('should')
const express = require('express')
const supertest = require('supertest')

const basicAuth = require('./index.js')

var app = express()

//Requires basic auth with username 'Admin' and password 'secret1234'
var staticUserAuth = basicAuth({
    users: {
        'Admin': 'secret1234'
    },
    challenge: false
})

//Uses a custom (synchronous) authorizer function
var customAuthorizerAuth = basicAuth({
    authorizer: myAuthorizer
})

//Uses a custom (synchronous) authorizer function
var customCompareAuth = basicAuth({
    authorizer: myComparingAuthorizer
})

//Same, but sends a basic auth challenge header when authorization fails
var challengeAuth = basicAuth({
    authorizer: myAuthorizer,
    challenge: true
})

//Uses a custom asynchronous authorizer function
var asyncAuth = basicAuth({
    authorizer: myAsyncAuthorizer,
    authorizeAsync: true
})

//Uses a custom response body function
var customBodyAuth = basicAuth({
    users: { 'Foo': 'bar' },
    unauthorizedResponse: getUnauthorizedResponse
})

//Uses a static response body
var staticBodyAuth = basicAuth({
    unauthorizedResponse: 'Haaaaaha'
})

//Uses a JSON response body
var jsonBodyAuth = basicAuth({
    unauthorizedResponse: { foo: 'bar' }
})

//Uses an unauthorized response middleware
var unauthorizedResponseMiddleware = basicAuth({
    unauthorizedResponse: function(req, res, next) {
        return res.status(200).send('This status code deceives you.');
    }
});

//Uses a custom realm
var realmAuth = basicAuth({
    challenge: true,
    realm: 'test'
})

//Uses a custom realm function
var realmFunctionAuth = basicAuth({
    challenge: true,
    realm: function (req) {
        return 'bla'
    }
})

app.get('/static', staticUserAuth, function(req, res) {
    res.status(200).send('You passed')
})

app.get('/custom', customAuthorizerAuth, function(req, res) {
    res.status(200).send('You passed')
})

app.get('/custom-compare', customCompareAuth, function(req, res) {
    res.status(200).send('You passed')
})

app.get('/challenge', challengeAuth, function(req, res) {
    res.status(200).send('You passed')
})

app.get('/async', asyncAuth, function(req, res) {
    res.status(200).send('You passed')
})

app.get('/custombody', customBodyAuth, function(req, res) {
    res.status(200).send('You passed')
})

app.get('/staticbody', staticBodyAuth, function(req, res) {
    res.status(200).send('You passed')
})

app.get('/jsonbody', jsonBodyAuth, function(req, res) {
    res.status(200).send('You passed')
})

app.get('/unauthorizedresponsemiddleware', unauthorizedResponseMiddleware, function(req, res) {
    res.status(200).send('You passed')
})

app.get('/realm', realmAuth, function(req, res) {
    res.status(200).send('You passed')
})

app.get('/realmfunction', realmFunctionAuth, function(req, res) {
    res.status(200).send('You passed')
})

//Custom authorizer checking if the username starts with 'A' and the password with 'secret'
function myAuthorizer(username, password) {
    return username.startsWith('A') && password.startsWith('secret')
}

//Same but asynchronous
function myAsyncAuthorizer(username, password, cb) {
    if(username.startsWith('A') && password.startsWith('secret'))
        return cb(null, true)
    else
        return cb(null, false)
}

function myComparingAuthorizer(username, password) {
    return basicAuth.safeCompare(username, 'Testeroni') & basicAuth.safeCompare(password, 'testsecret')
}

function getUnauthorizedResponse(req) {
    return req.auth ? ('Credentials ' + req.auth.user + ':' + req.auth.password + ' rejected') : 'No credentials provided'
}

describe('express-basic-auth', function() {
    describe('safe compare', function() {
        const safeCompare = basicAuth.safeCompare

        it('should return false on different inputs', function() {
            (!!safeCompare('asdf', 'rftghe')).should.be.false()
        })

        it('should return false on prefix inputs', function() {
            (!!safeCompare('some', 'something')).should.be.false()
        })

        it('should return false on different inputs', function() {
            (!!safeCompare('anothersecret', 'anothersecret')).should.be.true()
        })
    })

    describe('static users', function() {
        const endpoint = '/static'

        it('should reject on missing header', function(done) {
            supertest(app)
                .get(endpoint)
                .expect(401, done)
        })

        it('should reject on wrong credentials', function(done) {
            supertest(app)
                .get(endpoint)
                .auth('dude', 'stuff')
                .expect(401, done)
        })

        it('should reject on shorter prefix', function(done) {
            supertest(app)
                .get(endpoint)
                .auth('Admin', 'secret')
                .expect(401, done)
        })

        it('should reject without challenge', function(done) {
            supertest(app)
                .get(endpoint)
                .auth('dude', 'stuff')
                .expect(function (res) {
                    if(res.headers['WWW-Authenticate'])
                        throw new Error('Response should not have a challenge')
                })
                .expect(401, done)
        })

        it('should accept correct credentials', function(done) {
            supertest(app)
                .get(endpoint)
                .auth('Admin', 'secret1234')
                .expect(200, 'You passed', done)
        })
    })

    describe('custom authorizer', function() {
        const endpoint = '/custom'

        it('should reject on missing header', function(done) {
            supertest(app)
                .get(endpoint)
                .expect(401, done)
        })

        it('should reject on wrong credentials', function(done) {
            supertest(app)
                .get(endpoint)
                .auth('dude', 'stuff')
                .expect(401, done)
        })

        it('should accept fitting credentials', function(done) {
            supertest(app)
                .get(endpoint)
                .auth('Aloha', 'secretverymuch')
                .expect(200, 'You passed', done)
        })

        describe('with safe compare', function() {
            const endpoint = '/custom-compare'

            it('should reject wrong credentials', function(done) {
                supertest(app)
                    .get(endpoint)
                    .auth('bla', 'blub')
                    .expect(401, done)
            })

            it('should reject prefix credentials', function(done) {
                supertest(app)
                    .get(endpoint)
                    .auth('Test', 'test')
                    .expect(401, done)
            })

            it('should accept fitting credentials', function(done) {
                supertest(app)
                    .get(endpoint)
                    .auth('Testeroni', 'testsecret')
                    .expect(200, 'You passed', done)
            })
        })
    })

    describe('async authorizer', function() {
        const endpoint = '/async'

        it('should reject on missing header', function(done) {
            supertest(app)
                .get(endpoint)
                .expect(401, done)
        })

        it('should reject on wrong credentials', function(done) {
            supertest(app)
                .get(endpoint)
                .auth('dude', 'stuff')
                .expect(401, done)
        })

        it('should accept fitting credentials', function(done) {
            supertest(app)
                .get(endpoint)
                .auth('Aererer', 'secretiveStuff')
                .expect(200, 'You passed', done)
        })
    })

    describe('custom response body', function() {
        it('should reject on missing header and generate resposne message', function(done) {
            supertest(app)
                .get('/custombody')
                .expect(401, 'No credentials provided', done)
        })

        it('should reject on wrong credentials and generate response message', function(done) {
            supertest(app)
                .get('/custombody')
                .auth('dude', 'stuff')
                .expect(401, 'Credentials dude:stuff rejected', done)
        })

        it('should accept fitting credentials', function(done) {
            supertest(app)
                .get('/custombody')
                .auth('Foo', 'bar')
                .expect(200, 'You passed', done)
        })

        it('should reject and send static custom resposne message', function(done) {
            supertest(app)
            .get('/staticbody')
            .expect(401, 'Haaaaaha', done)
        })

        it('should reject and send static custom json resposne message', function(done) {
            supertest(app)
            .get('/jsonbody')
            .expect(401, { foo: 'bar' }, done)
        })

        it('should reject and use custom response middleware', function(done) {
            supertest(app)
            .get('/unauthorizedresponsemiddleware')
            .expect(200, 'This status code deceives you.', done)
        })
    })

    describe('challenge', function() {
        it('should reject with blank challenge', function(done) {
            supertest(app)
                .get('/challenge')
                .expect('WWW-Authenticate', 'Basic')
                .expect(401, done)
        })

        it('should reject with custom realm challenge', function(done) {
            supertest(app)
                .get('/realm')
                .expect('WWW-Authenticate', 'Basic realm="test"')
                .expect(401, done)
        })

        it('should reject with custom generated realm challenge', function(done) {
            supertest(app)
                .get('/realmfunction')
                .expect('WWW-Authenticate', 'Basic realm="bla"')
                .expect(401, done)
        })
    })
})


tests/test.js
require('should');

const express = require('express');
const supertest = require('supertest');

const basicAuth = require('../lib');

const app = express();

//Requires basic auth with username 'Admin' and password 'secret1234'
const staticUserAuth = basicAuth({
    users: {
        'Admin': 'secret1234'
    },
    challenge: false
});

//Uses a custom (synchronous) authorizer function
const customAuthorizerAuth = basicAuth({
    authorizer: myAuthorizer
});

//Uses a custom (synchronous) authorizer function
const customCompareAuth = basicAuth({
    authorizer: myComparingAuthorizer
});

//Same, but sends a basic auth challenge header when authorization fails
const challengeAuth = basicAuth({
    authorizer: myAuthorizer,
    challenge: true
});

//Uses a custom asynchronous authorizer function
const asyncAuth = basicAuth({
    authorizer: myAsyncAuthorizer,
    authorizeAsync: true
});

//Uses a custom response body function
const customBodyAuth = basicAuth({
    users: { 'Foo': 'bar' },
    unauthorizedResponse: getUnauthorizedResponse
});

//Uses a static response body
const staticBodyAuth = basicAuth({
    unauthorizedResponse: 'Haaaaaha'
});

//Uses a JSON response body
const jsonBodyAuth = basicAuth({
    unauthorizedResponse: { foo: 'bar' }
});

//Uses a custom realm
const realmAuth = basicAuth({
    challenge: true,
    realm: 'test'
});

//Uses a custom realm function
const realmFunctionAuth = basicAuth({
    challenge: true,
    realm: function () {
        return 'bla';
    }
});

app.get('/static', staticUserAuth, function(req, res) {
    res.status(200).send('You passed');
});

app.get('/custom', customAuthorizerAuth, function(req, res) {
    res.status(200).send('You passed');
});

app.get('/custom-compare', customCompareAuth, function(req, res) {
    res.status(200).send('You passed');
});

app.get('/challenge', challengeAuth, function(req, res) {
    res.status(200).send('You passed');
});

app.get('/async', asyncAuth, function(req, res) {
    res.status(200).send('You passed');
});

app.get('/custombody', customBodyAuth, function(req, res) {
    res.status(200).send('You passed');
});

app.get('/staticbody', staticBodyAuth, function(req, res) {
    res.status(200).send('You passed');
});

app.get('/jsonbody', jsonBodyAuth, function(req, res) {
    res.status(200).send('You passed');
});

app.get('/realm', realmAuth, function(req, res) {
    res.status(200).send('You passed');
});

app.get('/realmfunction', realmFunctionAuth, function(req, res) {
    res.status(200).send('You passed');
});

//Custom authorizer checking if the username starts with 'A' and the password with 'secret'
function myAuthorizer(username, password) {
    return username.startsWith('A') && password.startsWith('secret');
}

//Same but asynchronous
function myAsyncAuthorizer(username, password, cb) {
    if(username.startsWith('A') && password.startsWith('secret')) {
        return cb(null, true);
    } else {
        return cb(null, false);
    }
}

function myComparingAuthorizer(username, password) {
    return (
        basicAuth.safeCompare(username, 'Testeroni') &
        basicAuth.safeCompare(password, 'testsecret')
    );
}

function getUnauthorizedResponse(req) {
    return req.auth
        ? ('Credentials ' + req.auth.user + ':' + req.auth.password + ' rejected')
        : 'No credentials provided';
}

describe('express-basic-auth', function() {
    describe('safe compare', function() {
        const safeCompare = basicAuth.safeCompare;

        it('should return false on different inputs', function() {
            (!!safeCompare('asdf', 'rftghe')).should.be.false();
        });

        it('should return false on prefix inputs', function() {
            (!!safeCompare('some', 'something')).should.be.false();
        });

        it('should return false on different inputs', function() {
            (!!safeCompare('anothersecret', 'anothersecret')).should.be.true();
        });
    });

    describe('static users', function() {
        const endpoint = '/static';

        it('should reject on missing header', function(done) {
            supertest(app)
                .get(endpoint)
                .expect(401, done);
        });

        it('should reject on wrong credentials', function(done) {
            supertest(app)
                .get(endpoint)
                .auth('dude', 'stuff')
                .expect(401, done);
        });

        it('should reject on shorter prefix', function(done) {
            supertest(app)
                .get(endpoint)
                .auth('Admin', 'secret')
                .expect(401, done);
        });

        it('should reject without challenge', function(done) {
            supertest(app)
                .get(endpoint)
                .auth('dude', 'stuff')
                .expect(function (res) {
                    if(res.headers['WWW-Authenticate'])
                        throw new Error('Response should not have a challenge');
                })
                .expect(401, done);
        });

        it('should accept correct credentials', function(done) {
            supertest(app)
                .get(endpoint)
                .auth('Admin', 'secret1234')
                .expect(200, 'You passed', done);
        });
    });

    describe('custom authorizer', function() {
        const endpoint = '/custom';

        it('should reject on missing header', function(done) {
            supertest(app)
                .get(endpoint)
                .expect(401, done);
        });

        it('should reject on wrong credentials', function(done) {
            supertest(app)
                .get(endpoint)
                .auth('dude', 'stuff')
                .expect(401, done);
        });

        it('should accept fitting credentials', function(done) {
            supertest(app)
                .get(endpoint)
                .auth('Aloha', 'secretverymuch')
                .expect(200, 'You passed', done);
        });

        describe('with safe compare', function() {
            const endpoint = '/custom-compare';

            it('should reject wrong credentials', function(done) {
                supertest(app)
                    .get(endpoint)
                    .auth('bla', 'blub')
                    .expect(401, done);
            });

            it('should reject prefix credentials', function(done) {
                supertest(app)
                    .get(endpoint)
                    .auth('Test', 'test')
                    .expect(401, done);
            });

            it('should accept fitting credentials', function(done) {
                supertest(app)
                    .get(endpoint)
                    .auth('Testeroni', 'testsecret')
                    .expect(200, 'You passed', done);
            });
        });
    });

    describe('async authorizer', function() {
        const endpoint = '/async';

        it('should reject on missing header', function(done) {
            supertest(app)
                .get(endpoint)
                .expect(401, done);
        });

        it('should reject on wrong credentials', function(done) {
            supertest(app)
                .get(endpoint)
                .auth('dude', 'stuff')
                .expect(401, done);
        });

        it('should accept fitting credentials', function(done) {
            supertest(app)
                .get(endpoint)
                .auth('Aererer', 'secretiveStuff')
                .expect(200, 'You passed', done);
        });
    });

    describe('custom response body', function() {
        it('should reject on missing header and generate resposne message', function(done) {
            supertest(app)
                .get('/custombody')
                .expect(401, 'No credentials provided', done);
        });

        it('should reject on wrong credentials and generate response message', function(done) {
            supertest(app)
                .get('/custombody')
                .auth('dude', 'stuff')
                .expect(401, 'Credentials dude:stuff rejected', done);
        });

        it('should accept fitting credentials', function(done) {
            supertest(app)
                .get('/custombody')
                .auth('Foo', 'bar')
                .expect(200, 'You passed', done);
        });

        it('should reject and send static custom resposne message', function(done) {
            supertest(app)
                .get('/staticbody')
                .expect(401, 'Haaaaaha', done);
        });

        it('should reject and send static custom json resposne message', function(done) {
            supertest(app)
                .get('/jsonbody')
                .expect(401, { foo: 'bar' }, done);
        });
    });

    describe('challenge', function() {
        it('should reject with blank challenge', function(done) {
            supertest(app)
                .get('/challenge')
                .expect('WWW-Authenticate', 'Basic')
                .expect(401, done);
        });

        it('should reject with custom realm challenge', function(done) {
            supertest(app)
                .get('/realm')
                .expect('WWW-Authenticate', 'Basic realm="test"')
                .expect(401, done);
        });

        it('should reject with custom generated realm challenge', function(done) {
            supertest(app)
                .get('/realmfunction')
                .expect('WWW-Authenticate', 'Basic realm="bla"')
                .expect(401, done);
        });
    });
});


lib/example.js
const express = require('express');
const app = express();

const basicAuth = require('./index');

/**
* express-basic-auth
*
* Example server. Just run in the same folder:
*
* npm install express spresso-authy
*
* and then run this file with node ('node example.js')
*
* You can send GET requests to localhost:8080/async , /custom, /challenge or /static
* and see how it refuses or accepts your request matching the basic auth settings.
*/

//TODO: Implement some form of automatic testing against the example server

//Requires basic auth with username 'Admin' and password 'secret1234'
const staticUserAuth = basicAuth({
    users: {
        'Admin': 'secret1234'
    },
    challenge: false
});

//Uses a custom (synchronous) authorizer function
const customAuthorizerAuth = basicAuth({
    authorizer: myAuthorizer
});

//Same, but sends a basic auth challenge header when authorization fails
const challengeAuth = basicAuth({
    authorizer: myAuthorizer,
    challenge: true
});

//Uses a custom asynchronous authorizer function
const asyncAuth = basicAuth({
    authorizer: myAsyncAuthorizer,
    authorizeAsync: true
});

//Uses a custom response body function
const customBodyAuth = basicAuth({
    users: { 'Foo': 'bar' },
    unauthorizedResponse: getUnauthorizedResponse
});

//Uses a static response body
const staticBodyAuth = basicAuth({
    unauthorizedResponse: 'Haaaaaha'
});

//Uses a JSON response body
const jsonBodyAuth = basicAuth({
    unauthorizedResponse: { foo: 'bar' }
});

//Uses a custom realm
const realmAuth = basicAuth({
    challenge: true,
    realm: 'test'
});

//Uses a custom realm function
const realmFunctionAuth = basicAuth({
    challenge: true,
    realm: function () {
        return 'bla';
    }
});

app.get('/static', staticUserAuth, function(req, res) {
    res.status(200).send('You passed');
});

app.get('/custom', customAuthorizerAuth, function(req, res) {
    res.status(200).send('You passed');
});

app.get('/challenge', challengeAuth, function(req, res) {
    res.status(200).send('You passed');
});

app.get('/async', asyncAuth, function(req, res) {
    res.status(200).send('You passed');
});

app.get('/custombody', customBodyAuth, function(req, res) {
    res.status(200).send('You passed');
});

app.get('/staticbody', staticBodyAuth, function(req, res) {
    res.status(200).send('You passed');
});

app.get('/jsonbody', jsonBodyAuth, function(req, res) {
    res.status(200).send('You passed');
});

app.get('/realm', realmAuth, function(req, res) {
    res.status(200).send('You passed');
});

app.get('/realmfunction', realmFunctionAuth, function(req, res) {
    res.status(200).send('You passed');
});

app.listen(8080, function() {
    console.log('Listening!');
});

//Custom authorizer checking if the username starts with 'A' and the password with 'secret'
function myAuthorizer(username, password) {
    return username.startsWith('A') && password.startsWith('secret');
}

//Same but asynchronous
function myAsyncAuthorizer(username, password, cb) {
    if(username.startsWith('A') && password.startsWith('secret')) {
        return cb(null, true);
    } else {
        return cb(null, false);
    }
}

function getUnauthorizedResponse(req) {
    return req.auth
        ? ('Credentials ' + req.auth.user + ':' + req.auth.password + ' rejected')
        : 'No credentials provided';
}
